blocks/initialization.py

```diff
@@ -147,21 +148,39 @@ def __init__(self, mult=1):
         self.mult = mult
 
     def generate(self, rng, shape):
-        if len(shape) != 2:
-            raise ValueError
-        rows, cols = shape
-        return self.mult * numpy.eye(rows, cols, dtype=theano.config.floatX)
+        if len(shape) == 1:
+            eye_matrix = numpy.eye(1, shape[0], dtype=theano.config.floatX)
+            return self.mult * eye_matrix[0]
+        elif len(shape) == 2:
+            rows, cols = shape
+            eye_matrix = numpy.eye(rows, cols, dtype=theano.config.floatX)
+            return self.mult * eye_matrix
+        else:
+            raise ValueError("Identity initialization only works for 1D and"
+                             " 2D variables")
 
 
 class Orthogonal(NdarrayInitialization):
-    """Initialize a random orthogonal matrix.
+    r"""Initialize a random, approximately orthogonal matrix.
+
+    Only works for 1D and 2D variables.
 
-    Only works for 2D arrays.
+    When used to initialize a 1D parameter vector the elements will
+    be drawn i.i.d. from a normal distribution.
+
+    For 2D matrices we initialalize such that
+    :math:`M M^T \approx I` and :math:`M^T M \approx I` where
+    :math:`I` is an identity matrix of appropriate size.
 
     """
     def generate(self, rng, shape):
-        if len(shape) != 2:
-            raise ValueError
+        if len(shape) not in (1, 2):
+            raise ValueError("Orthogonal initialization only works for 1D and"
+                             " 2D variables")
+
+        if len(shape) == 1:
+            # Special case for 1D variables
+            return rng.randn(*shape).astype(theano.config.floatX)
```

Comment:
Although not as clearly defined, many people refer to a rectangular matrix where the columns are orthonormal as being an "orthogonal matrix". If you then consider a vector as a matrix with one row or column, it should have a dot product of one. The distinction of a dot product of a `1 x n` matrix with its transpose and the inner product of a vector of length `n` is a purely programmatic one; mathematically they are equivalent (and I don't like making the assumption that no one is going to compute dot products with vectors).

I understand the danger of having a badly initialized network, but shoehorning an initialization scheme to be mathematically incorrect based on an assumption of what the scheme will be used for is not the way to go. And it is just that, an assumption, because who is to say that someone doesn't want a vector that really is like a column from an orthogonal matrix?

I like @jbornschein's idea much better. This makes the initialization explicit to the user instead of being hidden away inside a magical initialization scheme.